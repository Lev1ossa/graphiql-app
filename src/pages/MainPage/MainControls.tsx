import { useLocale } from '@/hooks/useLocale'
import { useBoundStore } from '@/store'
import { Description } from '@mui/icons-material'
import { LoadingButton } from '@mui/lab'
import { Stack } from '@mui/material'

export default function MainControls() {
  const toggleAside = useBoundStore((s) => s.toggleAside)
  const { locale } = useLocale()

  const isSchemaFetching = useBoundStore((state) => state.isSchemaFetching)

  return (
    <Stack
      component="nav"
      py={1.5}
      spacing={1}
      direction="row"
      justifyContent="start"
    >
      <LoadingButton
        size="small"
        variant="contained"
        disableRipple
        onClick={() => toggleAside()}
        sx={{
          borderRadius: 0,
          paddingInline: 1.5,
        }}
        loading={isSchemaFetching}
        loadingPosition="start"
        startIcon={<Description />}
      >
        {locale.mainPage.button.openDocs}
      </LoadingButton>
    </Stack>
  )
}
