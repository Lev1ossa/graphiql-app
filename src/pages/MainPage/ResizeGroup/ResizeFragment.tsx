import { type ResizeFragmentProps } from './ResizeGroup'

export default function ResizeFragment(props: ResizeFragmentProps) {
  return props.children
}
