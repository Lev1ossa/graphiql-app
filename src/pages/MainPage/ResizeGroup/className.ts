export default {
  FragmentWindow: (index?: number) =>
    `ResizeGroup-Fragment-WindowBox${index ? `-${index}` : ''}`,
  FragmentResizer: (index?: number) =>
    `ResizeGroup-Fragment-SmartResizer${index ? `-${index}` : ''}`,
}
