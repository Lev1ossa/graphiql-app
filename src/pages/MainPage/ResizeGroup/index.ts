export { default as ResizeGroup } from './ResizeGroup'
export * from './Resizer'
