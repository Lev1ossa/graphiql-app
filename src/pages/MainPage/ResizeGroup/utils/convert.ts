export function toFr(value: number, share: number) {
  return value / share
}

export function toFrArray(values: number[]) {
  const share = values.reduce((s, v) => s + v, 0)
  return {
    fractions: values.map((v) => toFr(v, share)),
    share,
  }
}
