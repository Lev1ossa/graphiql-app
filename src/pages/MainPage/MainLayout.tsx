import { useBoundStore } from '@/store'
import { Box, Button, Drawer, useMediaQuery } from '@mui/material'
import { useLayoutEffect, useRef } from 'react'
import {
  ResizeGroup,
  ResizeGroupClassName,
  ResizerDefaults,
} from './ResizeGroup'
import ResizeFragment from './ResizeGroup/ResizeFragment'
import { type ResizeGroupProps } from './ResizeGroup/ResizeGroup'
import { TabGroupDefaults } from './TabGroup'
import TabGroup from './TabGroup/TabGroup'

type MainLayoutSlots = {
  documentation: React.ReactNode
  request: React.ReactNode
  response: React.ReactNode
  variables: React.ReactNode
  headers: React.ReactNode
}

const verticalLayoutStackProps = () => {
  return {
    stackProps: {
      sx: {
        [`& > .${ResizeGroupClassName.FragmentWindow(0)}`]: {
          minHeight:
            TabGroupDefaults.Tab.height - ResizerDefaults.padding + 'px',
        },
        [`& > .${ResizeGroupClassName.FragmentWindow(1)}`]: {
          minHeight: TabGroupDefaults.Tab.height + 'px',
        },
        [`& > .${ResizeGroupClassName.FragmentResizer()}`]: {
          marginBottom: -ResizerDefaults.padding + 'px',
        },
      },
    },
  } satisfies Pick<ResizeGroupProps, 'stackProps'>
}

const MainMobileLayout = ({
  documentation,
  headers,
  request,
  response,
  variables,
}: MainLayoutSlots) => {
  const isAsideOpen = useBoundStore((s) => s.isAsideOpen)
  const toggleAside = useBoundStore((s) => s.toggleAside)
  return (
    <Box height={1}>
      <Drawer
        open={isAsideOpen}
        onClose={() => toggleAside(false)}
        PaperProps={{
          sx: {
            width: 'clamp(320px, 100%, 500px)',
          },
        }}
      >
        <Button onClick={() => toggleAside(false)}>Close</Button>
        {documentation}
      </Drawer>
      <ResizeGroup
        direction="col"
        sizes={[0.7, 0.3]}
        preventUpdate={false}
        {...verticalLayoutStackProps()}
      >
        <ResizeFragment id="Row1">
          <Box height={1}>
            <TabGroup
              currentValue="Request"
              tabs={[
                { value: 'Request', jsx: request },
                { value: 'Response', jsx: response },
              ]}
            />
          </Box>
        </ResizeFragment>
        <ResizeFragment id="Row2">
          <Box height={1}>
            <TabGroup
              currentValue="Variables"
              tabs={[
                { value: 'Variables', jsx: variables },
                { value: 'Headers', jsx: headers },
              ]}
            />
          </Box>
        </ResizeFragment>
      </ResizeGroup>
    </Box>
  )
}

const MainDesktopLayout = ({
  documentation,
  headers,
  request,
  response,
  variables,
}: MainLayoutSlots) => {
  const isAsideOpen = useBoundStore((s) => s.isAsideOpen)
  return (
    <ResizeGroup direction="row" sizes={[0.2, 0.4, 0.4]} preventUpdate={false}>
      <ResizeFragment id="Col1" min={0.2} max={0.4} collapse={!isAsideOpen}>
        <Box height={1} overflow="auto">
          {documentation}
        </Box>
      </ResizeFragment>
      <ResizeFragment id="Col2" min={0.3}>
        <Box height={1}>
          <ResizeGroup
            direction="col"
            sizes={[0.7, 0.3]}
            preventUpdate={false}
            {...verticalLayoutStackProps()}
          >
            <ResizeFragment id="Col2Row1">
              <TabGroup
                currentValue="Request"
                tabs={[{ value: 'Request', jsx: request }]}
              />
            </ResizeFragment>
            <ResizeFragment id="Col2Row2">
              <TabGroup
                currentValue="Variables"
                tabs={[
                  { value: 'Variables', jsx: variables },
                  { value: 'Headers', jsx: headers },
                ]}
              />
            </ResizeFragment>
          </ResizeGroup>
        </Box>
      </ResizeFragment>
      <ResizeFragment id="Col3">
        <Box height={1}>
          <TabGroup
            currentValue="Response"
            tabs={[{ value: 'Response', jsx: response }]}
          />
        </Box>
      </ResizeFragment>
    </ResizeGroup>
  )
}

export default function MainLayout(slots: MainLayoutSlots) {
  const isMobile = useMediaQuery('(max-width: 600px)')
  const mainBoxRef = useRef<HTMLDivElement | null>(null)

  useLayoutEffect(() => {
    if (mainBoxRef.current) {
      const offsetTop = Math.round(mainBoxRef.current.offsetTop)
      const minHeight = 600
      const heightStr = `calc(max(100lvh - ${offsetTop + 'px'}, ${
        minHeight + 'px'
      }))`
      mainBoxRef.current.style.setProperty('height', heightStr)
    }
  }, [isMobile])

  return (
    <Box
      ref={mainBoxRef}
      sx={(theme) => ({
        borderWidth: ResizerDefaults.thickness + 'px',
        borderColor: theme.palette.divider,
        borderStyle: 'solid',
        '& ::-webkit-scrollbar': {
          height: 8 + 'px',
          width: 8 + 'px',
          overflow: 'overlay',
        },
        '& ::-webkit-scrollbar-thumb': {
          backgroundColor: theme.palette.divider,
        },
      })}
    >
      {isMobile ? (
        <MainMobileLayout {...slots} />
      ) : (
        <MainDesktopLayout {...slots} />
      )}
    </Box>
  )
}
