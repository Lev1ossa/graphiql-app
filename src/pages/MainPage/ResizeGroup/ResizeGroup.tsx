import { Box, Stack, type BoxProps, type StackProps } from '@mui/material'
import React, { Fragment, useRef } from 'react'
import { ResizerDefaults, SmartResizer } from './Resizer'
import { type ResizeEvent } from './Resizer/useResizer'
import className from './className'
import { moveFlexSpace, recalculateCollapse } from './utils/calculate'
import { toFr, toFrArray } from './utils/convert'

export type ResizeFragmentProps = {
  id: React.Key
  children: React.ReactNode
  min?: number
  max?: number
  collapse?: boolean
  resizerAfter?: boolean
  onSizeChange?: (prev: number, next: number, ref: HTMLDivElement) => void
  windowBoxProps?: BoxProps
  resizerBoxProps?: BoxProps
}
export type ResizeGroupChild = React.ReactElement<ResizeFragmentProps>
export type ResizeGroupProps = {
  children: React.ReactElement<ResizeFragmentProps>[]
  direction?: 'row' | 'col'
  sizes?: number[]
  onResize?: (fractions: number[]) => void
  preventUpdate?: boolean
  stackProps?: StackProps
}

export default function ResizeGroup(props: ResizeGroupProps) {
  const {
    direction = 'row',
    children,
    sizes = [],
    onResize = () => {},
    preventUpdate = true,
    stackProps,
  } = props

  const childrenProps = React.Children.map(
    children,
    (child: ResizeGroupChild) => {
      if (!Object.hasOwn(child.props, 'id')) {
        throw new Error(
          `Each <ResizeGroup /> child must have at least 'id' prop. Instead got '${
            child.type
          }' with: ${JSON.stringify(child.props)}`,
        )
      }
      const jsx = child.props.children
      const {
        id,
        collapse = [undefined, null, false, true].some((u) => u === jsx),
        min = 0,
        max = 1,
        onSizeChange = () => {},
        resizerAfter = true,
        windowBoxProps,
        resizerBoxProps,
      } = child.props
      return {
        children: jsx,
        id,
        collapse,
        min,
        max,
        onSizeChange,
        resizerAfter,
        windowBoxProps,
        resizerBoxProps,
      } satisfies ResizeFragmentProps
    },
  )

  const itemsRef = useRef<HTMLDivElement[]>([])

  const derivedSizes = recalculateCollapse({
    fractions: sizes,
    collapsed: childrenProps.map((ch) => ch.collapse),
  })

  const childrenSettings = childrenProps.map((cp, index) => ({
    ...cp,
    size: derivedSizes[index],
    minmax: [cp.min, cp.max] as [number, number],
    get ref() {
      return itemsRef.current[index]
    },
  }))

  const getChildArray = () => [...childrenSettings]
  const getChild = (index: number) => childrenSettings[index]

  const resizerOrientation = direction === 'row' ? 'vertical' : 'horizontal'
  const stackDirection = direction === 'row' ? 'row' : 'column'

  const getElSize = (el: HTMLElement) => {
    return direction === 'row' ? el.offsetWidth : el.offsetHeight
  }

  const resize = (nextFractions: number[]) => {
    const changeFractions = nextFractions.map((newSize, index) =>
      getChild(index).collapse ? getChild(index).size : newSize,
    )
    if (!preventUpdate) {
      getChildArray()
        .map((ch) => ch.ref)
        .forEach((el, index) => {
          const prev = getChild(index).size
          const next = changeFractions[index]
          el.style.setProperty('flex', `${changeFractions[index]} 1 0%`)
          if (prev !== next) {
            getChild(index).onSizeChange(prev, next, el)
          }
        })
    }
    onResize(changeFractions)
  }

  const onSmartResize = (e: ResizeEvent, dividerIndex: number) => {
    const currentItem = getChild(dividerIndex).ref
    if (currentItem) {
      const { fractions, share } = toFrArray(
        getChildArray()
          .map((ch) => ch.ref)
          .map(getElSize),
      )
      const value = (() => {
        const rect = currentItem.getBoundingClientRect()
        const value =
          direction === 'row'
            ? e.clientX - e.innerOffsetX - rect.left
            : e.clientY - e.innerOffsetY - rect.top
        return toFr(value, share)
      })()

      const nextSizes = moveFlexSpace({
        fractions,
        index: dividerIndex,
        value,
        minmax: getChildArray().map((ch) => ch.minmax),
      })

      if (nextSizes) resize(nextSizes)
    }
  }

  const display = (index: number) => ({
    display: getChild(index).collapse ? 'none' : undefined,
  })

  const edgeResizerPadding = (index: number) => {
    const padding = ResizerDefaults.padding + 'px'
    if (index === 0) {
      return {
        [direction === 'row' ? 'paddingLeft' : 'paddingTop']: padding,
      }
    }
    if (index === children.length - 2) {
      return {
        [direction === 'row' ? 'paddingRight' : 'paddingBottom']: padding,
      }
    }
  }

  const { FragmentWindow, FragmentResizer } = className
  const windowClassName = (i: number) => ({
    className: [FragmentWindow(), FragmentWindow(i)].join(' '),
  })
  const resizerClassName = (i: number) => ({
    className: [FragmentResizer(), FragmentResizer(i)].join(' '),
  })

  return (
    <Stack height={1} direction={stackDirection} {...stackProps}>
      {getChildArray().map((ch, index) => (
        <Fragment key={ch.id}>
          <Box
            overflow="hidden"
            style={{
              flex: `${ch.size} 1 0%`,
            }}
            ref={(node) => {
              itemsRef.current[index] = node as unknown as HTMLDivElement
            }}
            {...display(index)}
            {...windowClassName(index)}
            {...ch.windowBoxProps}
          >
            {ch.children}
          </Box>
          {ch.resizerAfter && index !== children.length - 1 && (
            <SmartResizer
              orientation={resizerOrientation}
              onResize={(e) => onSmartResize(e, index)}
              boxProps={{
                ...display(index),
                ...edgeResizerPadding(index),
                ...resizerClassName(index),
                ...ch.resizerBoxProps,
              }}
            />
          )}
        </Fragment>
      ))}
    </Stack>
  )
}
