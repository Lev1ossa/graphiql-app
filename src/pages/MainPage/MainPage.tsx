import { Footer } from '@/components/Footer/Footer'
import { Header } from '@/components/Header'
import { Response } from '@/components/Response'
import { Stack } from '@mui/material'
import MainControls from './MainControls'
import MainLayout from './MainLayout'

export default function MainPage() {
  return (
    <Stack>
      <Header />
      <MainControls />
      <MainLayout
        documentation={<div>Documentation</div>}
        request={<div>Request</div>}
        variables={<div>Variables</div>}
        headers={<div>Headers</div>}
        response={<Response />}
      />
      <Footer />
    </Stack>
  )
}
