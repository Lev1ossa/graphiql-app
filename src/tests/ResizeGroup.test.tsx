import { ResizeGroup, useResizer } from '@/pages/MainPage/ResizeGroup'
import ResizeFragment from '@/pages/MainPage/ResizeGroup/ResizeFragment'
import { describe, expect, test, vi } from 'vitest'
import renderInContext from './renderInContext'

import { moveFlexSpace } from '@/pages/MainPage/ResizeGroup/utils/calculate'
import { toFr, toFrArray } from '@/pages/MainPage/ResizeGroup/utils/convert'
import { fireEvent, render, renderHook, screen } from '@testing-library/react'

window.matchMedia = vi.fn().mockReturnValue(false)

describe('Correct ResizeGroup snapshot', () => {
  test('Correct ResizeGroup snapshot', () => {
    const { baseElement } = renderInContext(
      <ResizeGroup>
        <ResizeFragment id="0">{null}</ResizeFragment>
        <ResizeFragment id="1">{null}</ResizeFragment>
      </ResizeGroup>,
    )
    expect(baseElement).toMatchSnapshot()
  })
  test('Correct ResizeFragment snapshot', () => {
    const { baseElement } = renderInContext(
      <ResizeFragment id="0">{null}</ResizeFragment>,
    )
    expect(baseElement).toMatchSnapshot()
  })
})

describe('ResizeGroup core', () => {
  test('moveFlexSpace limited to min/max', () => {
    const moved = moveFlexSpace({
      fractions: [0.4, 0.2, 0.4],
      index: 0,
      sibling: 1,
      value: -0.3,
      minmax: [[0.1, 0.2]],
    })
    ;[0.1, 0.5, 0.4].forEach((e, i) => {
      expect(moved?.at(i)).toBeCloseTo(e)
    })
  })
  test('toFr/fromFr', () => {
    expect(toFr(100, 1000)).toBeCloseTo(1 / 10)
    expect(toFrArray([100, 200, 300])).toStrictEqual({
      fractions: [1 / 6, 2 / 6, 3 / 6],
      share: 600,
    })
  })
})

describe('ResizeGroup/useResizer', () => {
  test('Firing onStart/onEnd/onResize event', async () => {
    const onStart = vi.fn()
    const onEnd = vi.fn()
    const onResize = vi.fn()
    const {
      result: {
        current: { registerResizer },
      },
    } = renderHook(() => useResizer({ onResize, onStart, onEnd }))
    render(<span data-testid="divider" {...registerResizer()}></span>)

    const divider = screen.getByTestId('divider')

    fireEvent.pointerDown(divider)
    expect(onStart).toBeCalled()

    fireEvent.pointerMove(divider)
    expect(onResize).toBeCalled()

    fireEvent.pointerUp(divider)
    expect(onEnd).toBeCalled()
  })
})
