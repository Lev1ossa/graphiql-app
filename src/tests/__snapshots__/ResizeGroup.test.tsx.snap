// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Correct ResizeGroup snapshot > Correct ResizeFragment snapshot 1`] = `
<body>
  <div />
</body>
`;

exports[`Correct ResizeGroup snapshot > Correct ResizeGroup snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiStack-root css-88sa19-MuiStack-root"
    >
      <div
        class="ResizeGroup-Fragment-WindowBox ResizeGroup-Fragment-WindowBox MuiBox-root css-9dofl3"
      />
      <div
        class="ResizeGroup-Fragment-SmartResizer ResizeGroup-Fragment-SmartResizer MuiBox-root css-liov4t"
      >
        <hr
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-ldj3do-MuiDivider-root"
        />
      </div>
      <div
        class="ResizeGroup-Fragment-WindowBox ResizeGroup-Fragment-WindowBox-1 MuiBox-root css-9dofl3"
      />
    </div>
  </div>
</body>
`;
